File: LeetCodeRegular/search-insert-position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int n = nums.size();
        int left = 0, right = n-1, ind = n;
        while(left <= right){
            int mid = (left + right) / 2;
            if(target <= nums[mid]){
                ind = mid;
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }
        }
        return ind;   
    }
};

/*
二分查找法，关键在于不要把问题想得太复杂， 比如如果原数组没有需要插入的话，并不需要push_back这种命令，直接顺序放在最后，然后返回最后位置的索引就可以了
*/

File: LeetCodeRegular/算法/算法入门/LC_00035/search-insert-position.py
"""

评论 (2.2k)
题解 (3.2k)
提交记录
35. 搜索插入位置
给定一个排序数组和一个目标值，在数组中找到目标值，并返回其索引。如果目标值不存在于数组中，返回它将会被按顺序插入的位置。

你可以假设数组中无重复元素。

示例 1:

输入: [1,3,5,6], 5
输出: 2
示例 2:

输入: [1,3,5,6], 2
输出: 1
示例 3:

输入: [1,3,5,6], 7
输出: 4
示例 4:

输入: [1,3,5,6], 0
输出: 0

"""

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        n = len(nums)
        left, right = 0, n-1
        while left <= right:
            if nums[(left+right)//2] < target:
                left = (left+right)//2 + 1
            elif nums[(left+right)//2] > target:
                right = (left+right)//2 - 1
            else:
                return (left+right)//2
        if nums[n-1] < target:
                return n
        else:
            for i in range(n):
                if nums[i] < target:
                    i += 1
                else: 
                    return i


